References
----------

.. [Sokolov2016] Artem Sokolov, Daniel E. Carlin, Evan O. Paull, Robert Baertsch, Joshua M. Stuart (2016),
    *Pathway-Based Genomics Prediction using Generalized Elastic Net*,
    `PLOS Computational Biology <https://doi.org/10.1371/journal.pcbi.1004790>`__.

.. [Seninge2021] Lucas Seninge, Ioannis Anastopoulos, Hongxu Ding, Joshua Stuart (2021),
    *VEGA is an interpretable generative model for inferring biological network activity in single-cell transcriptomics*,
    `Nature Communications <https://doi.org/10.1038/s41467-021-26017-0>`__.

.. [Gayoso2022] Adam Gayoso*, Romain Lopez*, Galen Xing*, Pierre Boyeau, Valeh Valiollah Pour Amiri, Justin Hong, Katherine Wu, Michael Jayasuriya, Edouard Mehlman, Maxime Langevin, Yining Liu, Jules Samaran, Gabriel Misrachi, Achille Nazaret, Oscar Clivio, Chenling Xu, Tal Ashuach, Mohammad Lotfollahi, Valentine Svensson, Eduardo da Veiga Beltrame, Vitalii Kleshchevnikov, Carlos Talavera-Lopez, Lior Pachter, Fabian J Theis, Aaron Streets, Michael I Jordan, Jeffrey Regier, and Nir Yosef (2022),
   *A Python library for probabilistic analysis of single-cell omics data*,
   `Nature Biotechnology <http://dx.doi.org/10.1038/s41587-021-01206-w>`__.

.. [Lopez2018] Romain Lopez, Jeffrey Regier, Michael Cole, Michael I. Jordan, Nir Yosef (2018),
   *Deep generative modeling for single-cell transcriptomics*,
   `Nature Methods <https://www.nature.com/articles/s41592-018-0229-2.epdf?author_access_token=5sMbnZl1iBFitATlpKkddtRgN0jAjWel9jnR3ZoTv0P1-tTjoP-mBfrGiMqpQx63aBtxToJssRfpqQ482otMbBw2GIGGeinWV4cULBLPg4L4DpCg92dEtoMaB1crCRDG7DgtNrM_1j17VfvHfoy1cQ%3D%3D>`__.

.. [Rybakov2020] Sergei Rybakov, Mohammad Lotfollahi, Fabian J. Theis, F. Alexander Wolf (2021),
    *Learning interpretable latent autoencoder representations with annotations of feature sets*,
    `biorxiv <https://doi.org/10.1101/2020.12.02.401182>`__.

.. [Boyeau2019] Pierre Boyeau, Romain Lopez, Jeffrey Regier, Adam Gayoso, Michael I. Jordan, Nir Yosef (2019),
   *Deep generative models for detecting differential expression in single cells*,
   `Machine Learning in Computational Biology (MLCB) <https://www.biorxiv.org/content/biorxiv/early/2019/10/04/794289.full.pdf>`__.
